### PR TITLE
✨ [amp-analytics] Add Custom Browser Event Tracker

### PR DESCRIPTION
**Pull request by kalemuw**
This PR add a new feature to track a customizable set of triggers that are not currently supported by amp-analytics.
Comes in reponse to #17878 

Currently several events are defined to be tracked by the amp analytics component. However, as noted in the thread there are still several events that are not supported by amp-analytcis. This project begins the process of adding custom browser events and supports future needs for analytics data on events. 

The changes are currently added as an experiment in the PR.

Changes:
- In `events.js`
        - Add the custom browser event tracker for events blur and change
- In `test-custom-browser-events.js`
        - Add test for CBE tracker. 
        - Note: skipped over debounce to avoid waiting 
- In  `amp-cbe.amp.html`
        - Test page with two input fields and two dropdown menu. 
        - Tracks on-blur and on-change events on both elements. 

Reviewer: @dmanek @alanorozco

### CODE REVIEW

File: extensions/amp-analytics/0.1/events.js [hunk 1]
```diff
@@ -171,6 +185,14 @@ function isVideoTriggerType(triggerType) {
   return triggerType.startsWith('video');
 }
 
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isCustomBrowserTriggerType(triggerType) {
+  return Object.values(BrowserEventType).includes(triggerType);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:193]: `includes` is not supported in IE11. Use `indexOf`.

File: extensions/amp-analytics/amp-analytics.md [hunk 2]
```diff
@@ -546,7 +546,9 @@ each relevant trigger for more details.
 The selector properties are:
 
 -   `selector` This property is used to find an element or a collection of
-    elements using CSS/DOM query. The semantics of how the element is matched
+    elements using CSS/DOM query.
+
+- The semantics of how the element is matched
```
Comment (dmanek) [inline extensions/amp-analytics/amp-analytics.md:551]: Revert this change, no need for newlines.

File: extensions/amp-analytics/amp-analytics.md [hunk 3]
```diff
@@ -559,6 +561,24 @@ The selector properties are:
     of the `amp-analytics` tag that satisfies the given selector. The default
     value is `scope`.
 
+##### Selector Values <a name="selector-values"></a>
```
Comment (dmanek) [inline extensions/amp-analytics/amp-analytics.md:564]: Empty anchors `<a name="selector-values"></a>`?

**Comment by dmanek** [comment]
Revert and re-sync `package-lock.json`. It's causing the large diff in the PR.

File: extensions/amp-analytics/amp-analytics.md (hunk 2 above)
Comment (kalemuw) [reply extensions/amp-analytics/amp-analytics.md]: Just reverted the changes from amp-analytics.md and package-lock.json. 
Those were changes made by the amp-video multi selector documentation PR that has not been merged yet.

File: extensions/amp-analytics/0.1/events.js [hunk 4]
```diff
@@ -353,6 +353,12 @@ export class CustomBrowserEventTracker extends EventTracker {
 
   /** @override */
   dispose() {
+    const root = this.root.getRoot();
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:353]: Do you mean `getRootElement`?

Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js:354]: I used getRoot to keep the dispose function stylistically similar to the dispose function of the other trackers.

File: extensions/amp-analytics/0.1/events.js (hunk 1 above)
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Fixed!

File: extensions/amp-analytics/0.1/events.js [hunk 5]
```diff
@@ -171,6 +186,14 @@ function isVideoTriggerType(triggerType) {
   return triggerType.startsWith('video');
 }
 
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isCustomBrowserTriggerType(triggerType) {
+  return Object.values(BrowserEventType).indexOf(triggerType) > -1;
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:194]: `Object.values` is not support by IE. One option is to use `Object.keys` & then extract the values.

File: extensions/amp-analytics/0.1/events.js (hunk 4 above)
Comment (dmanek) [reply extensions/amp-analytics/0.1/events.js:354]: Nevermind. Type of `root` threw me off. It's an abstract class.

File: extensions/amp-analytics/0.1/events.js [hunk 6]
```diff
@@ -306,6 +332,83 @@ export class EventTracker {
...
+    const eventName = config['on'];
+
+    const targetReady = this.root.getElements(
+      context,
+      selector,
+      selectionMethod,
+      false
+    );
+
+    debounce(
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:384]: This is not the right way to use `debounce`. Couple of things you have wrong here:
1. You want to debounce the callback being passed in the `addEventListener`
2. `debounce` returns a function, so you typically want to use assign it to a var & pass into as an argument to `addEventListener`

File: extensions/amp-analytics/0.1/events.js [hunk 7]
```diff
@@ -306,6 +332,83 @@ export class EventTracker {
...
+    );
+
+    return this.observables_.add((event) => {
+      if (event.type !== eventName) {
+        return;
+      }
+      targetReady.then((targets) => {
+        targets.forEach((target) => {
+          const el = dev().assertElement(
+            event.target,
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:399]: You don't need to assert this. `event.target` should always be defined.

File: extensions/amp-analytics/0.1/test/test-custom-browser-events.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+          changeEventConfig,
+          () => {},
+          false
+        );
+
+        expect(tracker.observables_.getHandlerCount()).to.equal(1);
+
+        tracker.dispose();
+
+        expect(tracker.observables_).to.equal(null);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/test/test-custom-browser-events.js:104]: `to.be.null`

File: extensions/amp-analytics/0.1/test/test-custom-browser-events.js [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+        );
+        expect(() => {
+          tracker.add(analyticsElement, "change", config);
+        }).to.throw(
+          /Cannot have duplicate selectors in selectors list: #inputField,#inputField/
+        );
+      });
+
+      it('fires on one selector with on change', async () => {
+        const fn1 = env.sandbox.stub();
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/test/test-custom-browser-events.js:140]: rename instances of `fn1` to `listenerStub` or something more descriptive?

File: extensions/amp-analytics/0.1/test/test-custom-browser-events.js [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+        );
+
+        inputField.dispatchEvent(new Event('change'));
+
+        await macroTask();
+        expect(fn1).to.have.callCount(1);
+        expect(fn1).to.be.calledWith(
+          new AnalyticsEvent(inputField, 'change', {})
+        );
+        expect(getElementSpy).to.be.callCount(1);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/test/test-custom-browser-events.js:157]: Change instances of `to.be.callCount(1)` to `to.be.calledOnce`

File: extensions/amp-analytics/0.1/test/test-custom-browser-events.js [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+
+        inputField.dispatchEvent(new Event('blur'));
+
+        await macroTask();
+        expect(fn1).to.have.callCount(1);
+        expect(fn1).to.be.calledWith(
+          new AnalyticsEvent(inputField, 'blur', {})
+        );
+        expect(getElementSpy).to.be.callCount(1);
+        expect(tracker.observables_.handlers_.length).to.equal(1);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/test/test-custom-browser-events.js:180]: Typically private state of an object should not be verified in tests. But its ok to leave it for now.

File: extensions/amp-analytics/0.1/test/test-custom-browser-events.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+        expect(fn1.secondCall).to.be.calledWith(
+          new AnalyticsEvent(inputField2, 'blur', {})
+        );
+        expect(fn1.thirdCall).to.be.calledWith(
+          new AnalyticsEvent(inputField, 'change', {})
+        );
+        expect(fn1.lastCall).to.be.calledWith(
+          new AnalyticsEvent(inputField2, 'change', {})
+        );
+        expect(getElementSpy).to.be.callCount(2);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/test/test-custom-browser-events.js:274]: `.to.be.calledTwice`. Also its `to.have.callCount(...)`

File: extensions/amp-analytics/0.1/events.js [hunk 13]
```diff
@@ -306,6 +335,81 @@ export class EventTracker {
...
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
+    this.debouncedBoundOnSession_ = debounce(
+      this.root.ampdoc.win,
+      this.boundOnSession_,
+      500
```
Comment (erwinmombay) [inline extensions/amp-analytics/0.1/events.js:356]: lets hoist this value to the module level and add an explanation. thanks!

Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Sounds good. Just incorporated it in the latest commit

File: extensions/amp-analytics/0.1/events.js [hunk 14]
```diff
@@ -171,6 +187,17 @@ function isVideoTriggerType(triggerType) {
...
 }
 
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isCustomBrowserTriggerType(triggerType) {
+  const values = Object.keys(BrowserEventType).map((key) => {
+    return BrowserEventType[key];
+  });
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:197]: Don't do this. You're creating `values` every time this function is called. Instead create `values` outside of this function and assign to a var for faster lookups later.

File: extensions/amp-analytics/0.1/events.js [hunk 15]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
 
+/**
+ * Tracks custom browser events.
+ */
+export class CustomBrowserEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   */
+  constructor(root) {
+    const TIME_WAIT = 500;
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:347]: move this outside the class. Constants should never defined in a constructor.

File: extensions/amp-analytics/0.1/events.js [hunk 16]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+    super(root);
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:354]: Just curious, do u really need `bind` here?

File: extensions/amp-analytics/0.1/events.js [hunk 17]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+    this.debouncedBoundOnSession_ = debounce(
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:355]: Add type.

File: extensions/amp-analytics/0.1/events.js [hunk 18]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+    const targetReady = this.root.getElements(
+      context,
+      selector,
+      selectionMethod,
+      false
+    );
+
+    this.root
+      .getRootElement()
+      .addEventListener(eventName, this.debouncedBoundOnSession_, true);
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:401]: Remove 3rd param.

File: extensions/amp-analytics/0.1/events.js (hunk 16 above)
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js:354]: Yup. Fire by itself does not recieve the event unless bind is added.

File: extensions/amp-analytics/0.1/events.js [hunk 19]
```diff
@@ -171,6 +187,17 @@ function isVideoTriggerType(triggerType) {
...
+/**
+ * @param {string} triggerType
+ * @return {boolean}
+ */
+function isCustomBrowserTriggerType(triggerType) {
+  const values = Object.keys(BrowserEventType).map((key) => {
+    return BrowserEventType[key];
+  });
+  return values.indexOf(triggerType) > -1;
+}
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:199]: Like @dmanek mentioned, this is not the best way to do this. If you use `isEnumValue`, this will compile into the most efficient form.

```suggestion
import {isEnumValue} from '#core/types';

/**
 * @param {string} triggerType
 * @return {boolean}
 */
function isCustomBrowserTriggerType(triggerType) {
  return isEnumValue(BrowserEventType, triggerType);
}
```

File: extensions/amp-analytics/0.1/events.js (hunk 14 above)
Comment (alanorozco) [reply extensions/amp-analytics/0.1/events.js:197]: `isEnumValue` would be better, see [comment with suggestion.](https://github.com/ampproject/amphtml/pull/35193/files#r677585774)

File: extensions/amp-analytics/0.1/events.js (hunk 16 above)
Comment (alanorozco) [reply extensions/amp-analytics/0.1/events.js:354]: Re-referencing an instance method makes it lose its `this` binding otherwise.

File: extensions/amp-analytics/0.1/events.js [hunk 20]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+    userAssert(
+      selector.length,
+      'Missing required selector on browser event trigger'
+    );
+    assertUniqueSelectors(selector);
+
+    const selectionMethod = config['selectionMethod'] || null;
+    const eventName = config['on'];
+
+    const targetReady = this.root.getElements(
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:392]: nit: rename to `targetPromises` like on the video tracker.

File: extensions/amp-analytics/0.1/events.js [hunk 21]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+    this.boundOnSession_ = null;
+    this.observables_ = null;
+  }
+
+  /** @override */
+  add(context, eventType, config, listener) {
+    userAssert(
+      isExperimentOn(this.root.ampdoc.win, 'custom-browser-event-tracker'),
+      'expected global "custom-browser-event-tracker" experiment to be enabled'
+    );
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:377]: These experiment names are global, and it's not clear that this is about analytics. Let's rename.

```suggestion
    userAssert(
      isExperimentOn(this.root.ampdoc.win, 'analytics-browser-events'),
      'expected global "analytics-browser-events" experiment to be enabled'
    );
```

File: extensions/amp-analytics/0.1/events.js [hunk 22]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+      'Missing required selector on browser event trigger'
+    );
+    userAssert(
+      selector.length,
+      'Missing required selector on browser event trigger'
+    );
+    assertUniqueSelectors(selector);
+
+    const selectionMethod = config['selectionMethod'] || null;
+    const eventName = config['on'];
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:390]: 1. Let's not spill the concept of discerning by "browser events" into the user layer. Instead, let's specify the event itself on the user warnings.
2. Let's also use a single assertion for presence and length.
3. nit for style in `|| null` and destructure.

```suggestion
    const {
      'selectionMethod': selectionMethod = null,
      'on': eventName,
      'selector': selector,
    } = config;
    userAssert(
      selector?.length,
      `Missing required selector on configuration for trigger on "${on}"`
    );
    assertUniqueSelectors(selector);
```

File: extensions/amp-analytics/0.1/events.js [hunk 23]
```diff
@@ -306,6 +336,87 @@ export class EventTracker {
...
+      .addEventListener(eventName, this.debouncedBoundOnSession_, true);
+
+    return this.observables_.add((event) => {
+      if (event.type !== eventName) {
+        return;
+      }
+      targetReady.then((targets) => {
+        targets.forEach((target) => {
+          const el = event.target;
+          if (!target.contains(el)) {
```
[the remark this answers is not in the document]
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js:410]: Yes it does.

File: extensions/amp-analytics/0.1/events.js (hunk 14 above)
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js:197]: Sounds good. Thanks

File: extensions/amp-analytics/0.1/events.js (hunk 18 above)
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js:401]: removing it fails to add a listener for "blur" trigger.

File: extensions/amp-analytics/0.1/events.js [hunk 24]
```diff
@@ -306,6 +334,88 @@ export class EventTracker {
...
+   */
+  constructor(root) {
+    super(root);
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
+    /** @type {?function(!Event)} */ this.debouncedBoundOnSession_ = debounce(
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:352]: `/** @private {function()} */`
this.debouncedBoundOnSession_ = ...

File: extensions/amp-analytics/0.1/events.js (hunk 16 above)
Comment (dmanek) [reply extensions/amp-analytics/0.1/events.js:351]: Makes sense. My concern was its a bound function being passed to debounce, which makes it harder to debug & test.

**Review by dmanek — APPROVED** [review]

File: extensions/amp-analytics/0.1/events.js [hunk 25]
```diff
@@ -306,6 +334,90 @@ export class EventTracker {
...
+  constructor(root) {
+    super(root);
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
+
+    /** @private {?function(!Event)} */
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:353]:
```suggestion
    /** @private {!function(!Event):void} */
```

File: extensions/amp-analytics/0.1/events.js [hunk 26]
```diff
@@ -306,6 +334,90 @@ export class EventTracker {
...
+      context,
+      selector,
+      selectionMethod,
+      false
+    );
+
+    this.root
+      .getRootElement()
+      .addEventListener(eventName, this.debouncedBoundOnSession_, true);
+
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:403]: What happens if we have two `'on': 'blur' rules? That means that we would trigger `debouncedBoundOnSession_` twice, right? Could we keep track of which events we're already listening to?

Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Yes it triggers it twice. To avoid duplicate listener for the same event, I have made an object of the browser events with value initally set to false. Once a listener is added to the event, the value will be set to true. Because there is already a listener, further attempts to add a listener won't be successful. 
There's also an alternate solution of overriding `Element.addEventListener` to remember the added listeners, but I believe that the former is a better solution.

File: extensions/amp-analytics/0.1/events.js [hunk 27]
```diff
@@ -306,6 +348,94 @@ export class EventTracker {
...
+  /** @override */
+  dispose() {
+    const root = this.root.getRoot();
+    Object.keys(BrowserEventType).forEach((key) => {
+      root.removeEventListener(
+        BrowserEventType[key],
+        this.debouncedBoundOnSession_
+      );
+    });
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:384]: Since we would know the added listener types, let's remove those only.

```suggestion
    Object.keys(this.hasListener_).forEach((eventName) => {
      root.removeEventListener(
        eventName,
        this.debouncedBoundOnSession_
      );
    });
```

File: extensions/amp-analytics/0.1/events.js (hunk 26 above)
Comment (alanorozco) [reply extensions/amp-analytics/0.1/events.js]: Thanks for explaining. That solution sounds good to me, just made a few comments about its details.

File: extensions/amp-analytics/0.1/events.js [hunk 28]
```diff
@@ -306,6 +348,94 @@ export class EventTracker {
...
+      selectionMethod,
+      false
+    );
+
+    if (!status[eventName]) {
+      this.root
+        .getRootElement()
+        .addEventListener(eventName, this.debouncedBoundOnSession_, true);
+      status[eventName] = true;
+    }
```
[the remark this answers is not in the document]
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Good idea. I moved the function inside the class and initialized the eventListener tracker as a class property.

File: extensions/amp-analytics/0.1/events.js (hunk 27 above)
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Sounds good. Made the changes in the latest commit.

File: extensions/amp-analytics/0.1/events.js [hunk 29]
```diff
@@ -37,6 +39,8 @@ const DEFAULT_MAX_TIMER_LENGTH_SECONDS = 7200;
 const VARIABLE_DATA_ATTRIBUTE_KEY = /^vars(.+)/;
 const NO_UNLISTEN = function () {};
 const TAG = 'amp-analytics/events';
+const TIME_WAIT = 500;
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:42]: The name `TIME_WAIT` could apply to any sort of wait but it's used for something specific. Let's rename to `SESSION_DEBOUNCE_TIME_MS`.

File: extensions/amp-analytics/0.1/events.js [hunk 30]
```diff
@@ -306,6 +335,105 @@ export class EventTracker {
...
+    /** @private {?function(!Event)} */
+    this.boundOnSession_ = this.observables_.fire.bind(this.observables_);
+
+    /** @private {function(!Event):void} */
+    this.debouncedBoundOnSession_ = debounce(
+      this.root.ampdoc.win,
+      this.boundOnSession_,
+      TIME_WAIT
+    );
+    //browserEventListeners();
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:360]: Remove commented code.
```suggestion
    //browserEventListeners();
```

File: extensions/amp-analytics/0.1/events.js [hunk 31]
```diff
@@ -306,6 +334,93 @@ export class EventTracker {
...
+      if (event.type !== eventName) {
+        return;
+      }
+      targetPromises.then((targets) => {
+        targets.forEach((target) => {
+          const el = event.target;
+          if (!target.contains(el)) {
+            return;
+          }
+          listener(new AnalyticsEvent(target, eventName, event.detail));
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:417]: `event.detail` could potentially leak user information, let's set it to an empty object and add a TODO.

```suggestion
          // TODO(kalemuw): Allowlist properties from event.detail to pass as vars.
          listener(new AnalyticsEvent(target, eventName, {}));
```

File: extensions/amp-analytics/0.1/events.js [hunk 32]
```diff
@@ -1270,7 +1385,6 @@ export class VideoEventTracker extends EventTracker {
     this.boundOnSession_ = this.sessionObservable_.fire.bind(
       this.sessionObservable_
     );
-
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:1273]: nit: undo this line removal to keep the history clean.

**Review by alanorozco — CHANGES_REQUESTED** [review]

File: extensions/amp-analytics/0.1/events.js [hunk 33]
```diff
@@ -75,6 +84,14 @@ const TRACKER_TYPE = Object.freeze({
       return new ClickEventTracker(root);
     },
   },
+  [AnalyticsEventType.BROWSER_EVENT]: {
+    name: AnalyticsEventType.BROWSER_EVENT,
+    allowedFor: ALLOWED_FOR_ALL_ROOT_TYPES.concat(['timer']),
+    // Escape the temporal dead zone by not referencing a class directly.
+    klass: function (root) {
+      return new CustomBrowserEventTracker(root);
```
[the remark this answers is not in the document]
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Sounds good.

File: extensions/amp-analytics/0.1/events.js [hunk 34]
```diff
@@ -306,6 +334,93 @@ export class EventTracker {
...
+    );
+
+    const {
+      'on': eventName,
+      'selectionMethod': selectionMethod = null,
+      'selector': selector,
+    } = config;
+
+    userAssert(
+      selector.length,
```
[the remark this answers is not in the document]
Comment (kalemuw) [reply extensions/amp-analytics/0.1/events.js]: Thanks for the catch.

File: extensions/amp-analytics/0.1/events.js [hunk 35]
```diff
@@ -318,7 +429,7 @@ export class CustomEventTracker extends EventTracker {
 
     /** @const @private {!Object<string, !Observable<!AnalyticsEvent>>} */
     this.observables_ = {};
-
+
```
Comment (alanorozco) [inline extensions/amp-analytics/0.1/events.js:432]: I believe this won't pass the formatting check without this change.
```suggestion

```

File: tools/experiments/experiments-config.js [hunk 36]
```diff
@@ -64,6 +64,11 @@ export const EXPERIMENTS = [
     name: 'Allows the new lightbox experience to be used in A4A (prototype).',
     spec: 'https://github.com/ampproject/amphtml/issues/7743',
   },
+  {
+    id: 'analytics-browser-events',
+    name: 'Allows tracking of a custom set of browser events',
+    spec: 'https://github.com/ampproject/amphtml/pull/35193',
```
Comment (alanorozco) [inline tools/experiments/experiments-config.js:70]: Can we please add a `cleanupIssue`?

**Review by alanorozco — APPROVED** [review]

File: tools/experiments/experiments-config.js (hunk 36 above)
Comment (kalemuw) [reply tools/experiments/experiments-config.js:70]: Yup. Just added an issue: #35502

File: extensions/amp-analytics/0.1/events.js [hunk 37]
```diff
@@ -306,6 +334,89 @@ export class EventTracker {
...
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   */
+  constructor(root) {
+    super(root);
+
+    /** @private {?Observable<!Event>} */
+    this.observables_ = new Observable();
+
+    /** @private {?Object<BrowserEventType>} */
```
Comment (dmanek) [inline extensions/amp-analytics/0.1/events.js:350]: Sorry, my bad for the incorrect type earlier. Type should be `{!Object<BrowserEventType, boolean>}`.

**Review by dmanek — APPROVED** [review]

**Review by erwinmombay — APPROVED** [review]
